toolchain/parse/typed_nodes.h

```diff
@@ -0,0 +1,829 @@
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef CARBON_TOOLCHAIN_PARSE_TYPED_NODES_H_
+#define CARBON_TOOLCHAIN_PARSE_TYPED_NODES_H_
+
+#include "toolchain/parse/node_ids.h"
+#include "toolchain/parse/node_kind.h"
+
+namespace Carbon::Parse {
+
+// Helpers for defining different kinds of parse nodes.
+// ----------------------------------------------------
+
+// A pair of a list item and its optional following comma.
+template <typename Element, typename Comma>
+struct ListItem {
+  Element value;
+  std::optional<Comma> comma;
+};
+
+// A list of items, parameterized by the kind of the elements and comma.
+template <typename Element, typename Comma>
+using CommaSeparatedList = llvm::SmallVector<ListItem<Element, Comma>>;
+
+// This class provides a shorthand for defining parse node kinds for leaf nodes.
+template <const NodeKind& KindT, NodeCategory Category = NodeCategory::None>
+struct LeafNode {
+  static constexpr auto Kind = KindT.Define(Category);
+};
+
+// ----------------------------------------------------------------------------
+// Each type defined below corresponds to a parse node kind, and describes the
+// expected child structure of that parse node. Each should have a
+// `static constexpr Kind` member initialized to calling `Define` on the
+// corresponding `NodeKind`, in order to set its `NodeCategory`.
+//
+// This should be followed by a field for each direct child. These may use any
+// of these patterns (many of these are defined in `node_ids.h`):
+// - `NodeId` to match any single child node,
+// - `FooId` to require that child to have kind `Foo`,
+// - `AnyFooId` to require that child to have a kind in category `Foo`,
+// - `NodeIdOneOf<T, U>` to require the child to have kind `T::Kind` or
+//   `U::Kind`,
+// - `NodeIdNot<T>` to match any single child that doesn't have kind `T::Kind`,
+// - `llvm::SmallVector<T>` to match any number of children matching `T`
+// - `std::optional<T>` to match 0 or 1 children matching `T`
+// - `std::tuple<T...>` to match children matching `T...`
+// - `Aggregate`, for simple aggregate type `struct Aggregate { T x; U y; }`,
+//   to match children with types `T` and `U`.
+// ----------------------------------------------------------------------------
+
+// Error nodes
+// -----------
+
+// An invalid parse. Used to balance the parse tree. This type is here only to
+// ensure we have a type for each parse node kind. This node kind always has an
+// error, so can never be extracted.
+using InvalidParse =
+    LeafNode<NodeKind::InvalidParse, NodeCategory::Decl | NodeCategory::Expr>;
+
+// An invalid subtree. Always has an error so can never be extracted.
+using InvalidParseStart = LeafNode<NodeKind::InvalidParseStart>;
+struct InvalidParseSubtree {
+  static constexpr auto Kind =
+      NodeKind::InvalidParseSubtree.Define(NodeCategory::Decl);
+  InvalidParseStartId start;
+  llvm::SmallVector<NodeIdNot<InvalidParseStart>> extra;
```

Comment:
Could you add a blank line after the kind and the start of the fields? (Not suggesting the edit to all of them, but I can if needed.)

```suggestion
  static constexpr auto Kind =
      NodeKind::InvalidParseSubtree.Define(NodeCategory::Decl);
      
  InvalidParseStartId start;
  llvm::SmallVector<NodeIdNot<InvalidParseStart>> extra;
```